src/transformStateWithClones.js

```diff
@@ -6,8 +6,39 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(currentState, action.extraData);
+        stateHistory.push({ ...currentState });
+        break;
+
+      case 'removeProperties':
+        for (const ch of action.keysToRemove) {
+          if (currentState.hasOwnProperty(ch)) {
+            delete currentState[ch];
+          }
+        }
+        stateHistory.push({ ...currentState });
+        break;
+
+      case 'clear':
+        for (const j in currentState) {
+          if (currentState.hasOwnProperty(j)) {
+            delete currentState[j];
```

Comment:
Instead of deleting each property individually, you can simply assign an empty object to `currentState` to clear it. This is more efficient and concise.